### PR TITLE
Better way to configure unique terms within an expression

### PR DESCRIPTION
**Pull request by apmoriarty**
Push the configuration for the UniqueExpressionTermsVisitor down to the logic level vs. the DefaultQueryPlanner level.

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/config/LookupUUIDTune.java [hunk 1]
```diff
@@ -54,6 +54,8 @@ public void configure(BaseQueryLogic<Entry<Key,Value>> logic) {
             rsq.setSpeculativeScanning(speculativeScanning);
             rsq.setCacheModel(enableCaching);
             rsq.setPrimaryToSecondaryFieldMap(primaryToSecondaryFieldMap);
+            if (enforceUniqueTermsWithinExpressions)
+                rsq.setEnforceUniqueTermsWithinExpressions(true);
```
Comment (jwomeara) [inline warehouse/query-core/src/main/java/datawave/query/config/LookupUUIDTune.java]: Is the conditional here really necessary?

File: warehouse/query-core/src/main/java/datawave/query/config/LookupUUIDTune.java [hunk 2]
```diff
@@ -66,6 +68,9 @@ public void configure(BaseQueryLogic<Entry<Key,Value>> logic) {
                 SeekingQueryPlanner planner = new SeekingQueryPlanner();
                 planner.setMaxFieldHitsBeforeSeek(maxFieldHitsBeforeSeek);
                 planner.setMaxKeysBeforeSeek(maxKeysBeforeSeek);
+                if (enforceUniqueTermsWithinExpressions)
+                    planner.setEnforceUniqueTermsWithinExpressions(true);
```
Comment (jwomeara) [inline warehouse/query-core/src/main/java/datawave/query/config/LookupUUIDTune.java]: Same comment as above.

File: warehouse/query-core/src/main/java/datawave/query/config/LookupUUIDTune.java [hunk 3]
```diff
@@ -125,6 +130,8 @@ public void configure(GenericQueryConfiguration configuration) {
             if (maxShardsPerDayThreshold != -1) {
                 rsqc.setShardsPerDayThreshold(maxShardsPerDayThreshold);
             }
+            if (enforceUniqueTermsWithinExpressions)
+                rsqc.setEnforceUniqueTermsWithinExpressions(true);
```
Comment (jwomeara) [inline warehouse/query-core/src/main/java/datawave/query/config/LookupUUIDTune.java]: Same comment as above.

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java [hunk 4]
```diff
@@ -746,7 +746,8 @@ protected ASTJexlScript updateQueryTree(ScannerFactory scannerFactory, MetadataH
         
         stopwatch.stop();
         
-        if (enforceUniqueTermsWithinExpressions) {
+        // Enforce if either DefaultQueryPlanner or ShardQueryConfig is true
+        if (enforceUniqueTermsWithinExpressions || config.getEnforceUniqueTermsWithinExpressions()) {
```
Comment (jwomeara) [inline warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: Why does this term exist in two places?  Why not only have it in the shard query configuration?

Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: My understanding is that the LookupUUIDTune is used to override a select number of query configs, most of the options here are duplicates.

**Comment by apmoriarty** [comment]
> 
> 
> Why have this property in both places? I can guarantee you that's going to bite someone in the future when they think they've enabled/disabled it, but they missed the other config location. Also, why does it need to be moved to the shard config at all? Both the shard config and planner config are accessible in QueryLogicFactory.xml.

My understanding is that the LookupUUIDTune is used to override a select number of query configs, most of the options here are duplicates.

File: warehouse/query-core/src/main/java/datawave/query/config/LookupUUIDTune.java (hunk 1 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/config/LookupUUIDTune.java]: Nope, resolving in next commit.

File: warehouse/query-core/src/main/java/datawave/query/config/LookupUUIDTune.java (hunk 3 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/config/LookupUUIDTune.java]: Resolved

File: warehouse/query-core/src/main/java/datawave/query/config/LookupUUIDTune.java [hunk 5]
```diff
@@ -125,6 +125,8 @@ public void configure(GenericQueryConfiguration configuration) {
             if (maxShardsPerDayThreshold != -1) {
                 rsqc.setShardsPerDayThreshold(maxShardsPerDayThreshold);
             }
+            rsqc.setEnforceUniqueTermsWithinExpressions(enforceUniqueTermsWithinExpressions);
+
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/config/LookupUUIDTune.java]: we only need to set this once. If the value is stored in the BaseQueryLogic, no need to set it again in the ShardQueryConfiguration

Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/config/LookupUUIDTune.java]: I'll make that change, thanks

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java [hunk 6]
```diff
@@ -746,7 +740,8 @@ protected ASTJexlScript updateQueryTree(ScannerFactory scannerFactory, MetadataH
         
         stopwatch.stop();
         
-        if (enforceUniqueTermsWithinExpressions) {
+        // Enforce unique terms within an AND or OR expression.
+        if (config.getEnforceUniqueTermsWithinExpressions()) {
```
Comment (jwomeara) [inline warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: YES!

**Pull request merged**